src/components/shared/thread/ActivityGraphFills.js

```diff
@@ -11,37 +11,64 @@ import './ActivityGraph.css';
 import type {
   IndexIntoSamplesTable,
   IndexIntoCategoryList,
-  SamplesTable,
-  StackTable,
+  Thread,
 } from '../../../types/profile';
 import type {
   Milliseconds,
   DevicePixels,
   CssPixels,
 } from '../../../types/units';
-import type { ActivityGraphProps } from './ActivityGraph';
+
+/**
+ * Computing the ActivityGraph's fills requires a set of immutable properties that
+ * are used to compute the fills, but are never changed. In order to separate out
+ * the mutable values, collect all of these immutable settings in one object that
+ * can easily be shared between classes and functions.
+ */
+type LastRenderedGraphSettings = {|
+  +canvasPixelWidth: DevicePixels,
+  +canvasPixelHeight: DevicePixels,
+  +fullThread: Thread,
+  +interval: Milliseconds,
+  +rangeStart: Milliseconds,
+  +rangeEnd: Milliseconds,
+  +xPixelsPerMs: number,
+  +treeOrderSampleComparator: ?(
+    IndexIntoSamplesTable,
+    IndexIntoSamplesTable
+  ) => number,
+  +greyCategoryIndex: IndexIntoCategoryList,
+  +samplesSelectedStates: ?Array<boolean>,
+  +categoryDrawStyles: CategoryDrawStyles,
+|};
 
 type SampleContributionToPixel = {|
-  sample: IndexIntoSamplesTable,
-  contribution: number,
+  +sample: IndexIntoSamplesTable,
+  +contribution: number,
 |};
 
+/**
+ * The category fills are the computation that is ultimately returned for drawing
+ * the categories to the canvas. During the computation step, this value is mutated
+ * in place, but should be consumed immutably.
+ */
 type CategoryFill = {|
-  category: IndexIntoCategoryList,
-  perPixelContribution: Float32Array,
-  fillStyle: string | CanvasPattern,
+  +category: IndexIntoCategoryList,
+  +fillStyle: string | CanvasPattern,
+  // The Float32Array is mutated in place during the computation step.
+  +perPixelContribution: Float32Array,
```

Comment:
So the only mutation that I'm unhappy with at this point is the mutation of `perPixelContribution`, where we go from "this is how much this fill adds, per pixel" to "this is the shape of this fill's upper edge, after the fill has been stacked on the earlier fills in the fills array". Maybe things would be easier to understand if we kept those two things separate. The former can still be called `perPixelContribution`, and the latter can be called `accumulatedUpperEdge` maybe?

Reply:
Wow, that makes the code a lot easier to read and understand! I also broke out the accumulation into its own method.